### PR TITLE
parser: add charset validation (TiDB #4436)

### PR DESCRIPTION
**Pull request by wuudjac**
### What problem does this PR solve? 

Fix TiDB issue [#4436](https://github.com/pingcap/tidb/issues/4436), relates to TiDB PR [#9305](https://github.com/pingcap/tidb/pull/9305).

### What is changed and how it works?

If charset not found in `charset.charsetInfos`, return a `UnknownCharacterSet` error from parser.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Added charset validation in pattern `CharsetName` like what MySQL does, and use it in syntax
`CONVERT(expr USING transcoding_name)` and `CHAR(N USING charset_name)`.

Side effects

 - Modified `yyLexer` interface to return different types of error.

Related changes

 - Since `yyLexer` interface is modified, **bin/goyacc** may need rebuild in local develop environment by:
```
make clean
make all
```

### CODE REVIEW

**Comment by morgo** [comment]
@winkyao PTAL

**Comment by wuudjac** [comment]
To reviewers:

CI failed in TiDB because error type and error text changed, which does not fit **tidb** at branch `master`, for example:

> ... obtained string = "[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:1115]Unknown character set: 'gbk'"
> ... expected string = "Unknown charset gbk"

These test cases in TiDB is fixed in updated TiDB PR [#9305](https://github.com/pingcap/tidb/pull/9305), so the case above will become:
> ... obtained string = "[parser:1115]Unknown character set: 'gbk'"
> ... expected string = "[parser:1115]Unknown character set: 'gbk'"

With that said, currently the tests will pass only with **parser** at branch [`wuudjac:charset`](https://github.com/wuudjac/parser)  together with **tidb** at branch [`wuudjac:convert`](https://github.com/pingcap/tidb/pull/9305), could you please instruct me about the procedure of merging? Thank you.

And @lysu PTAL

File: go.mod1 [hunk 1]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/pingcap/errors v0.11.0
 	github.com/pingcap/tidb v0.0.0-20190218065808-69472bd1a6e9
 	github.com/pingcap/tipb v0.0.0-20190107072121-abbec73437b7
+	github.com/pkg/errors v0.8.1
```
Comment (tiancaiamao) [inline go.mod1]: You can use github.com/pincap/errors which is forked from github.com/pkg/errors @wuudjac

**Comment by tiancaiamao** [comment]
@winkyao PTAL

File: go.mod1 (hunk 1 above)
Comment (wuudjac) [reply go.mod1]: @tiancaiamao It's my mistake, fixed in [`f969710`](https://github.com/pingcap/parser/pull/220/commits/f96971034f46cad81282625d6c1d210e3a0a5650).

**Comment by tiancaiamao** [comment]
PTAL @lysu @XuHuaiyu @zz-jason

**Comment by lysu** [comment]
ping @winkyao @zz-jason

**Comment by zz-jason** [comment]
@wuudjac sorry for the late reply, this PR looks good to me, could you merge the master branch and resolve conflicts?

**Comment by wuudjac** [comment]
@zz-jason conflict resolved, but I don't know how will the merge happen, as the tests are fix in TiDB PR [#9305](https://github.com/pingcap/tidb/pull/9305).

Do I need to remove `replace` in TiDB PR [#9305](https://github.com/pingcap/tidb/pull/9305) and update **go.mod**, **go.sum** to follow latest pingcap/parser master, after this PR is somehow merged?

> To reviewers:
> 
> CI failed in TiDB because error type and error text changed, which does not fit **tidb** at branch `master`, for example:
> 
> > ... obtained string = "[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:1115]Unknown character set: 'gbk'"
> > ... expected string = "Unknown charset gbk"
> 
> These test cases in TiDB is fixed in updated TiDB PR [#9305](https://github.com/pingcap/tidb/pull/9305), so the case above will become:
> 
> > ... obtained string = "[parser:1115]Unknown character set: 'gbk'"
> > ... expected string = "[parser:1115]Unknown character set: 'gbk'"
> 
> With that said, currently the tests will pass only with **parser** at branch [`wuudjac:charset`](https://github.com/wuudjac/parser) together with **tidb** at branch [`wuudjac:convert`](https://github.com/pingcap/tidb/pull/9305), could you please instruct me about the procedure of merging? Thank you.
> 
> And @lysu PTAL

**Pull request merged**